rosbag2/include/rosbag2/compression_interfaces/base_compressor_interface.hpp

```diff
@@ -0,0 +1,76 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2__COMPRESSION_INTERFACES__BASE_COMPRESSOR_INTERFACE_HPP_
+#define ROSBAG2__COMPRESSION_INTERFACES__BASE_COMPRESSOR_INTERFACE_HPP_
+
+#include <memory>
+#include <string>
+
+#include "rosbag2/visibility_control.hpp"
+
+/**
+ * An interface for developers adding a new compression algorithm to rosbag2. These functions
+ * must be implemented to that a writer can properly compress a file or bag message.
+ * A corresponding decompressor must also be implemented.
+ *
+ * Example file compression usage:
+ *
+ * MyCompressor my_compressor();
+ * std::string current_uri = storage.get_relative_path();
+ * std::string compressed_path_uri = my_compressor.compress_uri(current_uri);
+ * relative_file_paths.push_back(compressed_path_uri);
+ *
+ * Example message compression usage:
+ *
+ * MyCompressor my_compressor();
+ * std::shared_ptr<SerializedBagMessage> bag_message = std::make_shared<SerializedBagMessage>();
+ * ...fill message
+ * std::shared_ptr<SerializedBagMessage> compressed_message =
+ *   my_compressor.compress_serialized_bag_message(bag_message);
```

Comment:
That compression implementation lives outside the `rosbag2` package, right? I mean the use case you describe here is applicable independently and outside the realm of rosbag.

Reply:
Implementation of compression can be outside of rosbag2. However, we want to have a default one that can be used out of the box with the package.
We can remove the compression implementation out of the rosbag2 namespace.

Reply:
The namespace doesn't matter, really, but the dependencies kind of do. I would propose to shift the implementation into its own package (even if default) which could be designed similar to the `default_converter_plugins` or `default_storage_plugins`.

Reply:
Sounds good, I can make that change.
Will make a `default_compression_plugin` package and start by adding the necessary compression features in the reader (another PR) so you can see how it works.